Additional request around notify-keyspace-events to support Spring Data

**Is your feature request related to a problem? Please describe.**
Spring Data Redis repository likes to set config values and when it can't it bombs out on start.

https://github.com/spring-projects/spring-data-redis/issues/2670

```
Caused by: io.lettuce.core.RedisCommandExecutionException: ERR Unknown option or number of arguments for CONFIG SET - 'notify-keyspace-events'
    at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:147) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:116) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.AsyncCommand.completeResult(AsyncCommand.java:120) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.AsyncCommand.complete(AsyncCommand.java:111) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.CommandWrapper.complete(CommandWrapper.java:63) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.CommandHandler.complete(CommandHandler.java:745) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.CommandHandler.decode(CommandHandler.java:680) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.lettuce.core.protocol.CommandHandler.channelRead(CommandHandler.java:597) ~[lettuce-core-6.3.2.RELEASE.jar!/:6.3.2.RELEASE/8941aea]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994) ~[netty-common-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.110.Final.jar!/:4.1.110.Final]
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.110.Final.jar!/:4.1.110.Final]
    at java.base/java.lang.Thread.run(Thread.java:840) ~[na:na]
```

I'm trying to use Redis (and in this case DragonflyDB) as a shared session store for a high availability SpringBoot based application.

**Describe the solution you'd like**
I'd like to be able to use Spring Data Redis with DragonflyDB.

**Describe alternatives you've considered**
Keep using Redis sentinel and become increasingly despondent with the experience in EKS.

**Additional context**

Related to https://github.com/dragonflydb/dragonfly/issues/1657

https://www.shinyproxy.io/documentation/configuration/#session-persistence-and-aws-elasticache - upstream configuration details for elasticache that disables this by default so explicit configuration is needed there.

_Using DragonflyDB 1.22.1 via the Operator with 3 replicas._

--- a/src/server/db_slice.cc
+++ b/src/server/db_slice.cc
@@ -1451,6 +1451,15 @@ void DbSlice::ResetEvents() {
   events_ = {};
 }
 
+void DbSlice::SetNotifyKeyspaceEvents(std::string_view notify_keyspace_events) {
+  if (notify_keyspace_events.empty()) {
+    for (auto& db : db_arr_) {
+      db->expired_keys_events_.clear();
+    }
+  }
+  expired_keys_events_recording_ = !notify_keyspace_events.empty();
+}
+
 void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
   if (client_tracking_map_.empty())
     return;

--- a/src/server/db_slice.cc
+++ b/src/server/db_slice.cc
@@ -1452,11 +1452,6 @@ void DbSlice::ResetEvents() {
 }
 
 void DbSlice::SetNotifyKeyspaceEvents(std::string_view notify_keyspace_events) {
-  if (notify_keyspace_events.empty()) {
-    for (auto& db : db_arr_) {
-      db->expired_keys_events_.clear();
-    }
-  }
   expired_keys_events_recording_ = !notify_keyspace_events.empty();
 }
 

--- a/src/server/db_slice.h
+++ b/src/server/db_slice.h
@@ -511,6 +511,10 @@ class DbSlice {
     return pt->Traverse(cursor, std::forward<Cb>(cb));
   }
 
+  // Does not check for non supported events. Callers must parse the string and reject it
+  // if it's not empty and not EX.
+  void SetNotifyKeyspaceEvents(std::string_view notify_keyspace_events);
+
  private:
   void PreUpdate(DbIndex db_ind, Iterator it, std::string_view key);
   void PostUpdate(DbIndex db_ind, Iterator it, std::string_view key, size_t orig_size);

--- a/src/server/main_service.cc
+++ b/src/server/main_service.cc
@@ -931,6 +931,25 @@ void Service::Init(util::AcceptServer* acceptor, std::vector<facade::Listener*> 
   config_registry.RegisterMutable("table_growth_margin");
   config_registry.RegisterMutable("tcp_keepalive");
 
+  config_registry.RegisterMutable(
+      "notify_keyspace_events", [pool = &pp_](const absl::CommandLineFlag& flag) {
+        auto res = flag.TryGet<std::string>();
+        if (!res.has_value() || (!res->empty() && !absl::EqualsIgnoreCase(*res, "EX"))) {
+          return false;
+        }
+
+        pool->AwaitBrief([&res](unsigned, auto*) {
+          auto* shard = EngineShard::tlocal();
+          if (shard) {
+            auto shard_id = shard->shard_id();
+            auto& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard_id);
+            db_slice.SetNotifyKeyspaceEvents(*res);
+          }
+        });
+
+        return true;
+      });
+
   serialization_max_chunk_size = GetFlag(FLAGS_serialization_max_chunk_size);
   uint32_t shard_num = GetFlag(FLAGS_num_shards);
   if (shard_num == 0 || shard_num > pp_.size()) {

--- a/tests/dragonfly/connection_test.py
+++ b/tests/dragonfly/connection_test.py
@@ -411,20 +411,15 @@ async def test_subscribers_with_active_publisher(df_server: DflyInstance, max_co
     await async_pool.disconnect()
 
 
-@dfly_args({"notify_keyspace_events": "Ex"})
-async def test_keyspace_events(async_client: aioredis.Redis):
-    pclient = async_client.pubsub()
-    await pclient.subscribe("__keyevent@0__:expired")
-
+async def produce_expiring_keys(async_client: aioredis.Redis):
     keys = []
     for i in range(10, 50):
         keys.append(f"k{i}")
         await async_client.set(keys[-1], "X", px=200 + i * 10)
+    return keys
 
-    # We don't support immediate expiration:
-    # keys += ['immediate']
-    # await async_client.set(keys[-1], 'Y', exat=123) # expired 50 years ago
 
+async def collect_expiring_events(pclient, keys):
     events = []
     async for message in pclient.listen():
         if message["type"] == "subscribe":
@@ -433,8 +428,48 @@ async def test_keyspace_events(async_client: aioredis.Redis):
         events.append(message)
         if len(events) >= len(keys):
             break
+    return events
+
+
+@dfly_args({"notify_keyspace_events": "Ex"})
+async def test_keyspace_events(async_client: aioredis.Redis):
+    pclient = async_client.pubsub()
+    await pclient.subscribe("__keyevent@0__:expired")
+
+    keys = await produce_expiring_keys(async_client)
+
+    # We don't support immediate expiration:
+    # keys += ['immediate']
+    # await async_client.set(keys[-1], 'Y', exat=123) # expired 50 years ago
+
+    events = await collect_expiring_events(pclient, keys)
 
     assert set(ev["data"] for ev in events) == set(keys)
+
+
+async def test_keyspace_events_config_set(async_client: aioredis.Redis):
+    # nonsense does not make sense as argument, we only accept ex or empty string
+    with pytest.raises((ResponseError)):
+        await async_client.config_set("notify_keyspace_events", "nonsense")
+
+    await async_client.config_set("notify_keyspace_events", "ex")
+    pclient = async_client.pubsub()
+    await pclient.subscribe("__keyevent@0__:expired")
+
+    keys = await produce_expiring_keys(async_client)
+
+    events = await collect_expiring_events(pclient, keys)
+
+    assert set(ev["data"] for ev in events) == set(keys)
+
+    keys = await produce_expiring_keys(async_client)
+    await async_client.config_set("notify_keyspace_events", "")
+    try:
+        async with async_timeout.timeout(1):
+            await collect_expiring_events(pclient, keys)
+        assert False
+    except:
+        pass
 
 
 async def test_big_command(df_server, size=8 * 1024):

